concordium-consensus/src/Concordium/Scheduler/Environment.hs

```diff
@@ -682,7 +692,7 @@ defaultSuccess wtc = \ls events -> do
 
 {-# INLINE liftLocal #-}
 liftLocal :: Monad m => m a -> LocalT r m a
-liftLocal m = LocalT (ContT (\k -> RWST (\r s -> m >>= \f -> runRWST (k f) r s)))
+liftLocal m = LocalT (ContT (\k -> (lift . lift) m >>= k))
```

Comment:
The original reason for the more complicated expression was that it would use the bind of the inner monad `m`, rather than of the `RWST`.

That said, I don't know what gets optimized better, because presumably GHC will inline the `lift`s and the `bind`s of the state and reader monads, which should simplify quite well as a result.

Have you tried to see which version works better?

Reply:
I checked and using the inner monad seems to be slightly faster.